core/state: set-based journalling

holiman opened the pull request:
This is a second attempt at https://github.com/ethereum/go-ethereum/pull/30500 . 

This PR introduces set-based journalling, where the journalling-events are basically stored in per-scoped maps.

Whenever we enter a new call scope, we create a new scoped journal. Changes within the same scoped journal overwrite eachother. For example: if a scope updates a balance to from 6 to 5 then 4, then 3, there will only be one preimage in the journal. As opposed to the old journal (linear_journal), which would have multiple entries: [ prev: 6, prev: 5, prev:4].

The linear / appending journal is wasteful on memory, and also slow on rollbacks, since each change is rolled back individually.


------------------


@karalabe reminded me of the burntpix benchmark, on which this PR excels, so I thought it was worth another chance. 

`master`:
```
[user@work go-ethereum]$ ./evm_master run --prestate ./burntpix.json --receiver 0x49206861766520746F6F206D7563682074696D65  --input 0xa4de9ab4000000000000000000000000000000000000000000000000000000000F1FD58E000000000000000000000000000000000000000000000000000000000007A120 --bench | tail -c +131 | sed 's/[0]*$//' | xxd -r -p > output.svg
```
```
EVM gas used:    5642735088
execution time:  49.349209526s
allocations:     915684
allocated bytes: 175333368
```
This PR: 
```
[user@work go-ethereum]$ ./evm_setjournal run --prestate ./burntpix.json --receiver 0x49206861766520746F6F206D7563682074696D65  --input 0xa4de9ab4000000000000000000000000000000000000000000000000000000000F1FD58E000000000000000000000000000000000000000000000000000000000007A120 --bench | tail -c +131 | sed 's/[0]*$//' | xxd -r -p > output.svg
```
```
EVM gas used:    5642735088
execution time:  48.740463584s
allocations:     30198
allocated bytes: 30308272
```
Allocations, `915K` -> `30K`,
Allocated bytes: `175M` -> `30M`

jwasinger commented on core/genesis.go:
```diff
@@ -152,6 +152,7 @@ func flushAlloc(ga *types.GenesisAlloc, triedb *triedb.Database) (common.Hash, e
 	if err != nil {
 		return common.Hash{}, err
 	}
+
```
rm

holiman replied:
```suggestion
```

jwasinger commented on core/state/setjournal.go:
```diff
@@ -0,0 +1,486 @@
+	if parent.refund == -1 {
+		parent.refund = j.refund
+	}
+	// Revert changes to accounts
```
I think you meant to write `merge` instead of `revert` in a lot of places in this method.

holiman replied:
Thanks, fixed!

karalabe commented on core/state/journal_api.go:
```diff
@@ -0,0 +1,86 @@
+	// DiscardSnapshot removes the snapshot with the given id; after calling this
+	// method, it is no longer possible to revert to that particular snapshot, the
+	// changes are considered part of the parent scope.
+	DiscardSnapshot(revid int)
```
Also, suggest moving this up below revertSnapshot, to make the triplet (snapshot, revert, discard) groupped together.

karalabe commented on core/state/journal_api.go:
```diff
@@ -0,0 +1,86 @@
+	snapshot() int
+
+	// revertToSnapshot reverts all state changes made since the given revision.
+	revertToSnapshot(revid int, s *StateDB)
```
revertTOsnapshot and discardSnapshot seems asymmetric (one discards something, the other reverts **up to** something).

Would it make sense to unify these? I.e. have discard also take the parent until which to discard (if revert knows where to revert, discard should also know? shoudln't it be called in the same spot?).

Also, discard doesn't really need an id, does it, you will only ever discard the head?

holiman replied:
Myeah, I guess you are right. At the moment, though, the `linear_journal` (a.k.a our current journal), doesn't actually do anything on discard: it expects the revert to simply tell it where to go. 

So yeah, for the sake of clarity and semantic cleanliness, I will change that.

holiman replied:
Right. I have a comment in `journal_set.go`: 
```
	want := len(j.entries) - 1
	have := id
	if want != have {
		if want == 0 && id == 1 {
			// If a transaction is applied successfully, the statedb.Finalize will
			// end by clearing and resetting the journal. Invoking a discardSnapshot
			// afterwards will lead us here.
			// Let's not panic, but it's ok to complain a bit
			log.Error("Extraneous invocation to discard snapshot")
			return
		} else {
			panic(fmt.Sprintf("journalling error, want discard(%d), have discard(%d)", want, have))
		}
	}
```
So, yes, it does expect `discard` only to occur on the last entry. However, it's because the statedb.Finalize does this: 
```
	// Invalidate journal because reverting across transactions is not allowed.
	s.clearJournalAndRefund()
```
So, we apply a transaction. If it succeeds, the statedb will `clear` the journal, making it new again. But, the miner does this: 
```
	var (
		snap = env.state.Snapshot()
		gp   = env.gasPool.Gas()
	)
	receipt, err := core.ApplyTransaction(miner.chainConfig, miner.chain, &env.coinbase, env.gasPool, env.state, env.header, tx, &env.header.GasUsed, vm.Config{})
	if err != nil {
		env.state.RevertToSnapshot(snap)
		env.gasPool.SetGas(gp)
	}
```
So, the miner will potentially call `RevertToSnapshot`. It will _never_ call `Discard`, since that happens internally inside `ApplyTransaction`. From an APi-perspective, this is bad. The `Snapshot()` and `RevertToSnapshot` are collected in one place, but the `Discard` is not.

karalabe reviewed (changes requested):
Random nits for now

holiman commented:
I tried to depoy this on `bench06` without wiping it. cc @rjl493456442 I guess you were running a future db format on it? Hope I didn't break any important run.. ?
```
Nov 07 13:25:36 bench06.ethdevops.io geth INFO [11-07|12:25:36.509] Using pebble as the backing database
Nov 07 13:25:36 bench06.ethdevops.io geth INFO [11-07|12:25:36.509] Allocated cache and file handles database=/datadir/geth/geth/chaindata cache=2.00GiB handles=524,288
Nov 07 13:25:39 bench06.ethdevops.io geth Chain metadata
Nov 07 13:25:39 bench06.ethdevops.io geth databaseVersion: 9 (0x9)
Nov 07 13:25:39 bench06.ethdevops.io geth headBlockHash: 0xe637caa29a2821cffeafd8db1906e765b0be7e4d8350123e156d757d3db62b2f
Nov 07 13:25:39 bench06.ethdevops.io geth headFastBlockHash: 0xe637caa29a2821cffeafd8db1906e765b0be7e4d8350123e156d757d3db62b2f
Nov 07 13:25:39 bench06.ethdevops.io geth headHeaderHash: 0xe637caa29a2821cffeafd8db1906e765b0be7e4d8350123e156d757d3db62b2f
Nov 07 13:25:39 bench06.ethdevops.io geth lastPivotNumber: 20977634 (0x14017e2)
Nov 07 13:25:39 bench06.ethdevops.io geth len(snapshotSyncStatus): 285 bytes
Nov 07 13:25:39 bench06.ethdevops.io geth snapshotDisabled: false
Nov 07 13:25:39 bench06.ethdevops.io geth snapshotJournal: 7938463 bytes
Nov 07 13:25:39 bench06.ethdevops.io geth snapshotRecoveryNumber: <nil>
Nov 07 13:25:39 bench06.ethdevops.io geth snapshotRoot: 0x047f51ef93470331dee55a068d094b27ccf97dc6e5ce104ae99587bfefb8ff12
Nov 07 13:25:39 bench06.ethdevops.io geth txIndexTail: 18785755 (0x11ea5db)
Nov 07 13:25:39 bench06.ethdevops.io geth
SkeletonSyncStatus: {"Subchains":[{"Head":21135734,"Tail":21135734,"Next":"0x36399eb652517a6a8618352d09c8d1cdfdf63a5bda68e3dc4a20c5eb290a27a8"}],"Finalized":21050420}
Nov 07 13:25:40 bench06.ethdevops.io geth Fatal: Failed to register the Ethereum service: rlp: input list has too many elements for rawdb.freezerTableMeta
Nov 07 13:25:40 bench06.ethdevops.io geth Fatal: Failed to register the Ethereum service: rlp: input list has too many elements for rawdb.freezerTableMeta

```

holiman commented:
Deploying now, so 
- `bench05`: this PR, but using `linear_journal` 
- `bench06`: this PR, but using `setjournal`

karalabe commented:
I was running Gary’s pr

On Thu, 7 Nov 2024 at 13:31, Martin HS ***@***.***> wrote:

> Deploying now, so
>
>    - bench05: this PR, but using linear_journal
>    - bench06: this PR, but using setjournal
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ethereum/go-ethereum/pull/30660#issuecomment-2462119232>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAA7UGN6TTGHA7JCOLAD2IDZ7NMRRAVCNFSM6AAAAABQOC4ON6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDINRSGEYTSMRTGI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

holiman commented:
> Deploying now, so
> 
>     * `bench05`: this PR, but using `linear_journal`
> 
>     * `bench06`: this PR, but using `setjournal`

Has been running fine for ~5 days. Nothing noteworthy with regards do differences in charts on the two machines.

rjl493456442 commented:
> I tried to depoy this on `bench06` without wiping it. cc @rjl493456442 I guess you were running a future db format on it? Hope I didn't break any important run.. ?
> 
> ```
> Nov 07 13:25:36 bench06.ethdevops.io geth INFO [11-07|12:25:36.509] Using pebble as the backing database
> Nov 07 13:25:36 bench06.ethdevops.io geth INFO [11-07|12:25:36.509] Allocated cache and file handles database=/datadir/geth/geth/chaindata cache=2.00GiB handles=524,288
> Nov 07 13:25:39 bench06.ethdevops.io geth Chain metadata
> Nov 07 13:25:39 bench06.ethdevops.io geth databaseVersion: 9 (0x9)
> Nov 07 13:25:39 bench06.ethdevops.io geth headBlockHash: 0xe637caa29a2821cffeafd8db1906e765b0be7e4d8350123e156d757d3db62b2f
> Nov 07 13:25:39 bench06.ethdevops.io geth headFastBlockHash: 0xe637caa29a2821cffeafd8db1906e765b0be7e4d8350123e156d757d3db62b2f
> Nov 07 13:25:39 bench06.ethdevops.io geth headHeaderHash: 0xe637caa29a2821cffeafd8db1906e765b0be7e4d8350123e156d757d3db62b2f
> Nov 07 13:25:39 bench06.ethdevops.io geth lastPivotNumber: 20977634 (0x14017e2)
> Nov 07 13:25:39 bench06.ethdevops.io geth len(snapshotSyncStatus): 285 bytes
> Nov 07 13:25:39 bench06.ethdevops.io geth snapshotDisabled: false
> Nov 07 13:25:39 bench06.ethdevops.io geth snapshotJournal: 7938463 bytes
> Nov 07 13:25:39 bench06.ethdevops.io geth snapshotRecoveryNumber: <nil>
> Nov 07 13:25:39 bench06.ethdevops.io geth snapshotRoot: 0x047f51ef93470331dee55a068d094b27ccf97dc6e5ce104ae99587bfefb8ff12
> Nov 07 13:25:39 bench06.ethdevops.io geth txIndexTail: 18785755 (0x11ea5db)
> Nov 07 13:25:39 bench06.ethdevops.io geth
> SkeletonSyncStatus: {"Subchains":[{"Head":21135734,"Tail":21135734,"Next":"0x36399eb652517a6a8618352d09c8d1cdfdf63a5bda68e3dc4a20c5eb290a27a8"}],"Finalized":21050420}
> Nov 07 13:25:40 bench06.ethdevops.io geth Fatal: Failed to register the Ethereum service: rlp: input list has too many elements for rawdb.freezerTableMeta
> Nov 07 13:25:40 bench06.ethdevops.io geth Fatal: Failed to register the Ethereum service: rlp: input list has too many elements for rawdb.freezerTableMeta
> ```

Yes, the db version is 9, so it's confirmed!

MariusVanDerWijden commented on core/state/journal.go:
```diff
@@ -1,4 +1,4 @@
-// Copyright 2016 The go-ethereum Authors
+// Copyright 2024 the go-ethereum Authors
```
```suggestion
// Copyright 2024 The go-ethereum Authors
```
Don't know what to do here, should we keep the original date or the new one. I think the new one is okay, since the content of the file is significantly changes.

MariusVanDerWijden commented on core/state/journal_linear.go:
```diff
@@ -0,0 +1,500 @@
+
+// revertSnapshot reverts all state changes made since the last call to snapshot().
+func (j *linearJournal) revertSnapshot(s *StateDB) {
+	id := len(j.revisions) - 1
```
Same here, this will panic if there was no previous call to snapshot()

MariusVanDerWijden replied:
Interestingly enough the set based journaling will not panic on revertSnapshot or discardSnapshot

MariusVanDerWijden commented on core/state/journal_set.go:
```diff
@@ -0,0 +1,470 @@
+
+// revertSnapshot reverts all state changes made since the last call to snapshot().
+func (j *sparseJournal) revertSnapshot(s *StateDB) {
+	id := len(j.entries) - 1
```
If you call revertSnapshot on the top-most snapshot, you will end up with a broken journal, since the initial (hidden) snapshot will be reverted. I don't think its intended behavior, since all subsequent calls will panics

MariusVanDerWijden replied:
A fix would be to call `s.snapshot()` if `id == 0`

MariusVanDerWijden commented:
I built a little fuzzer for differential fuzzing the two journals against each other here: https://github.com/MariusVanDerWijden/go-ethereum/tree/journal_on_heapfix_fuzz, its in the first commit, I also added some things on top, so that the fuzzer doesn't panic

holiman commented:
> We could add it to this PR, but please make it so that it's deterministic

I've done this now

holiman commented:
Rebased on master, which changed with eip-7702 so the journal-api now does not implicitly require the pre-code to be nil. 

This has not yet been implemented in the set-based journal, but I have added it to the fuzzer, so the failure is detected. It was harder than I thought to detect, because it's a bit sneaky: if a journal sets back the codehash but not the code, then the changes pass fine: because `code` is only an in-memory byteslice, the codehash is the consensus object. 

Hence, a broken journal would subtly corrupt the in-memory stateobject. This can be detected by checking the codehash against the code. 

This is not yet fixed in set-journal, because I want to think a bit on how to best handle that.

holiman commented:
> if a journal sets back the codehash but not the code, then the changes pass fine: because `code` is only an in-memory byteslice, the codehash is the consensus object.

Actually, there may be nothing wrong with this. 
```
	code []byte // contract bytecode, which gets set when code is loaded
```
It should be fine to set the `code` to `nil`, while resetting the previous `codeHash`. With the assumption that

1. Code will be loaded on-demand when needed
2. Code already exists in the code-database. 

The latter assumption is interesting. Is there any case where we can undo a setCode, and revert to an older code, and the older code is _not_ stored in the db?
